libraries/AP_Devo_Telem/AP_Devo_Telem.cpp

```diff
@@ -0,0 +1,128 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/* 
+   DEVO Telemetry library
+*/
+
+#define DEVOM_SYNC_BYTE        0xAA
+
+
+#define AP_SERIALMANAGER_DEVO_TELEM_BAUD        38400
+#define AP_SERIALMANAGER_DEVO_BUFSIZE_RX        0
+#define AP_SERIALMANAGER_DEVO_BUFSIZE_TX        32
+
+#include "AP_Devo_Telem.h"
+
+extern const AP_HAL::HAL& hal;
+
+
+//constructor
+AP_DEVO_Telem::AP_DEVO_Telem(AP_AHRS &ahrs) :
+    _ahrs(ahrs),
+    _battery(AP::battery())
+{
+    devoPacket.header = DEVOM_SYNC_BYTE;    
+}
+
+// init - perform require initialisation including detecting which protocol to use
+void AP_DEVO_Telem::init(const AP_SerialManager& serial_manager)
+{
+    // check for DEVO_DPort
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Devo_Telem, 0))) {
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        // initialise uart
+        _port->begin(AP_SERIALMANAGER_DEVO_TELEM_BAUD, AP_SERIALMANAGER_DEVO_BUFSIZE_RX, AP_SERIALMANAGER_DEVO_BUFSIZE_TX);
+
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_DEVO_Telem::tick, void));
+    }
+}
+
+
+uint32_t AP_DEVO_Telem::gpsDdToDmsFormat(float ddm){
+    int32_t deg = (int32_t)ddm;
+    float mm = (ddm - deg) * 60.0f;
+
+    mm = ((float)deg * 100.0f + mm) /100.0;
+
+    if ((mm < -180.0f) || (mm > 180.0f)){
+        mm = 0.0f;
+    }
+
+    return mm * 1.0e7;
+}
+
+
+/*
+  send_frames - sends updates down telemetry link 
+  should be called by main program at 1hz
+*/
+
+#define DEVO_SPEED_FACTOR 0.0194384f
+
+void AP_DEVO_Telem::send_frames(uint8_t control_mode)
+{
+    // return immediately if not initialised
+    if (_port == nullptr) {
+        return;
+    }
+
+    const AP_GPS &gps = AP::gps();
+
+    if (gps.status() >= (enum AP_GPS::GPS_Status)GPS_FIX_TYPE_3D_FIX) {
+        struct Location loc = gps.location();
+
+        devoPacket.lat = gpsDdToDmsFormat(loc.lat);
+        devoPacket.lon = gpsDdToDmsFormat(loc.lng);
+        devoPacket.speed = (int16_t)(gps.ground_speed() * DEVO_SPEED_FACTOR * 100.0f);  // * 100 for cm
+
+        /*
+          Note that this isn't actually barometric altitude, it is the
+          inertial nav estimate of altitude above home.
+        */
+        float alt;
+        _ahrs.get_relative_position_D_home(alt);
+        alt *= -100.0f;
+        devoPacket.alt   = alt * -100.0f; // NED, in cm!
```

Comment:
The AHRS method returns it in NED in meters. So the math here doesn't make sense.

Reply:
this code was from your offer somewhere in this discussion, comment also from your above post. Conclusions?

Reply:
Code is perfectly fine, it's the comment (that is yours, not mine) that is wrong. `devoPacket.alt` will have altitude in U in cm, which is usually what users want, so it should be correct - but it's not what the comment says.

Reply:
comment is about coordinates in Ardupilot is in NED so we should to reverse sign. Seems that there is the word "was" required in this comment

Reply:
Comments like that usually say what will be in the assigned variable, not what in one of the operands. But the biggest issue with the comment is it's inconsistency: if it's about the `alt` variable, then it should say *NED, in m!*...

Reply:
I understood your question. usually the author is the last one to whom you should apply for comments - for him so everything is clear. But now the comment is corrected to reflect the actual state of affairs.